framework/src/auxkernels/BoundsAuxBase.C

```diff
@@ -33,24 +34,59 @@ BoundsAuxBase::BoundsAuxBase(const InputParameters & parameters)
     _bounded_var_name(parameters.get<NonlinearVariableName>("bounded_variable")),
     _var(_subproblem.getStandardVariable(_tid, _bounded_var_name))
 {
-  if (!isNodal())
-    mooseError("BoundsAuxBase must be used on a nodal auxiliary variable!");
   if (!Moose::PetscSupport::isSNESVI(*dynamic_cast<FEProblemBase *>(&_subproblem)))
     mooseDoOnce(mooseWarning(
         "A variational inequalities solver must be used in conjunction with BoundsAux"));
+
+  // Check that the bounded variable is of a supported type
+  if (!_var.isNodal() && (_var.feType().order != CONSTANT))
+    paramError("bounded_variable", "Bounded variable must be nodal or of a CONSTANT order!");
+
+  auto & dummy =
+      _aux_sys.getActualFieldVariable<Real>(_tid, parameters.get<AuxVariableName>("variable"));
+
+  // Check that the dummy variable matches the bounded variable
+  if (dummy.feType() != _var.feType())
+    paramError("variable",
+               "Dummy bound aux. variable and bounded variable must use the same finite element "
+               "order and family");
 }
 
 Real
 BoundsAuxBase::computeValue()
 {
-  if (_current_node->n_dofs(_nl_sys.number(), _bounded_var.number()) > 0)
-  {
-    // The zero is for the component, this will only work for Lagrange variables!
-    dof_id_type dof = _current_node->dof_number(_nl_sys.number(), _bounded_var.number(), 0);
+  dof_id_type dof = getDoFIndex();
 
+  if (dof != std::numeric_limits<dof_id_type>::max())
```

Comment:
I think this condition will never be true based on the logic below. So maybe change this to an assertion.

Reply:
I think it s a block restriction safeguard. This preserves the old behavior

Reply:
@hugary1995 it is possible to be true. Think of a first order Lagrange variable on a second order mesh. There will be no dofs on non-vertex nodes

Reply:
I see. @GiudGiud can you add the reasoning as comments in the source code?

Reply:
sure